### ISSUE TITLE
Why can't you perform additional operations on the loss function under tf.gradienttape?

### ISSUE THREAD

**Issue by 1148330040** [tensorflow/tensorflow#51927]
tf：2.2
env：linux
this is a **Triplet extraction model**
subject_labels is normal:   (batch_size,  max_len,  2)
object_labels is : (batch_size, max_len, len(predicate), 2)
For object_ Model I need to overwrite the padding part and only modify the text part。

However, after I perform the overwrite operation, although the obtained l**oss value** has no problem in the form, it is put into tf.gradienttape, and the weight value of Bert model cannot be obtained

When my loss function is as follows，Only TF related APIs are used:
`
def loss4model(t, p, l, obj=False):
    if obj:
        loss_value = tf.keras.losses.binary_crossentropy(y_true=t, y_pred=p)
        loss_value = tf.reduce_mean(loss_value, axis=2)
    else:
        loss_value = tf.keras.losses.binary_crossentropy(y_true=t, y_pred=p)

    loss_value = tf.reduce_mean(loss_value)

    return loss_value
`

When the model performs the following operations（in the tf.GradientTape）, it is normal and the loss function is normal
![1631267664(1)](https://user-images.githubusercontent.com/34124260/132836145-3ab467d2-e009-467d-9d9e-ce0570f06b47.png)
like this：
![1631268153(1)](https://user-images.githubusercontent.com/34124260/132837171-dcf8277e-f279-4ca1-ad41-32ffe796ad03.png)

But when I need to do some additional operations on the loss function, for example
`
def loss4model(t, p, l, obj=False):
    """I need to overwrite the padding part and only lose the text par
    """
    if obj:
        loss_value = tf.keras.losses.binary_crossentropy(y_true=t, y_pred=p)
        loss_value = tf.reduce_mean(loss_value, axis=2)
    else:
        loss_value = tf.keras.losses.binary_crossentropy(y_true=t, y_pred=p)
        
    loss_values = []
    for num, weights in enumerate(loss_value):
       # Get the loss value within the specified range through the length of TF. Gather() and len(text)
        indices = np.arange(l[num])
        loss = tf.gather(weights, indices)
        loss = tf.reduce_mean(loss)
        loss_values.append(loss.numpy())

    loss_value = tf.reduce_mean(tf.constant(loss_values))

    return loss_value
`
![1631267758(1)](https://user-images.githubusercontent.com/34124260/132836279-2dcc5c49-85d4-4f5e-af6c-88cf6ad8dd28.png)

You can see that the **loss value is a normal tensor**, but I **can't get the weight value** of the model as above，the weight is None：
![1631267805(1)](https://user-images.githubusercontent.com/34124260/132836401-95a9f696-49a8-4f02-9866-3c2b2b3bb27e.png)

I tried to define a loss directly in the model llike this
![1631268227(1)](https://user-images.githubusercontent.com/34124260/132837321-8134abbc-5b26-4520-aad0-a0604ab703c1.png)
it cant help
![1631268249(1)](https://user-images.githubusercontent.com/34124260/132837364-1dca5332-3abb-4ad5-90bf-de9f7ad79aba.png)

so, i want to know, loss_value is just a tf.tensor, Why can it affect the weight update and acquisition of the model?
and How can I still make loss work after I operate it?

**Comment by sushreebarsa** [comment]
@1148330040 
Can you please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new/choose),In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

**Comment by 1148330040** [comment]
> @1148330040
> Can you please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new/choose),In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thanks!

code  It's a little long, so I'll describe it as much as possible. I'd like to ask you

**Comment by sushreebarsa** [comment]
@1148330040 It looks like you are using an older Version of Tensorflow (2.2). Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version (2.6.0) and let us know if the issue still persists? Please have a look at the[ link](https://www.tensorflow.org/api_docs/python/tf/GradientTape) & let us know if it helps? Thanks!
